test/test_hydrostatic_free_surface_models.jl

```diff
@@ -97,12 +97,39 @@ topos_3d = ((Periodic, Periodic, Bounded),
             for arch in archs, FT in float_types
                 grid = RectilinearGrid(arch, FT, topology=topo, size=(1, 1, 1), extent=(1, 2, 3))
                 model = HydrostaticFreeSurfaceModel(grid=grid)
-
                 @test model isa HydrostaticFreeSurfaceModel
             end
         end
     end
 
+    @testset "Halo size check in model constructor" begin
+        for topo in topos_3d
+            grid = RectilinearGrid(topology=topo, size=(1, 1, 1), extent=(1, 2, 3))
+            hcabd_closure = HorizontallyCurvilinearAnisotropicBiharmonicDiffusivity()
+
+            @test_throws ArgumentError HydrostaticFreeSurfaceModel(grid=grid, tracer_advection=CenteredFourthOrder())
+            @test_throws ArgumentError HydrostaticFreeSurfaceModel(grid=grid, tracer_advection=UpwindBiasedThirdOrder())
+            @test_throws ArgumentError HydrostaticFreeSurfaceModel(grid=grid, tracer_advection=UpwindBiasedFifthOrder())
+            @test_throws ArgumentError HydrostaticFreeSurfaceModel(grid=grid, momentum_advection=UpwindBiasedFifthOrder())
+            @test_throws ArgumentError HydrostaticFreeSurfaceModel(grid=grid, closure=hcabd_closure)
```

Comment:
I don't think that

https://github.com/CliMA/Oceananigans.jl/blob/a72479f6895d80d0c297ae8576b4044160f4511b/src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_model.jl#L108-L114

will throw an error here. 

Why is the halo size not adequate here? Because `HorizontallyCurvilinearAnisotropicBiharmonicDiffusivity()` requires 2 halos? I can't figure out where the resizing/validation occurs for this case.

Reply:
However, something is not right. The warning seems to be claiming that halo size was bumped up to (2, 2, 2). But:

```Julia
julia> model = HydrostaticFreeSurfaceModel(grid=grid, closure=hcabd_closure)
┌ Warning: HydrostaticFreeSurfaceModel is experimental. Use with caution!
└ @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/Research/OC2/src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_model.jl:106
┌ Warning: Inflating model grid halo size to (2, 2, 2) and recreating grid. The model grid will be different from the input grid. To avoid this warning, pass halo=(2, 2, 2) when constructing the grid.
└ @ Oceananigans.Grids ~/Research/OC2/src/Grids/automatic_halo_sizing.jl:41
┌ Warning: Inflating model grid halo size to (2, 2, 2) and recreating grid. The model grid will be different from the input grid. To avoid this warning, pass halo=(2, 2, 2) when constructing the grid.
└ @ Oceananigans.Grids ~/Research/OC2/src/Grids/automatic_halo_sizing.jl:41
HydrostaticFreeSurfaceModel{CPU, Float64}(time = 0 seconds, iteration = 0)
├── grid: RectilinearGrid{Float64, Periodic, Bounded, Bounded}(Nx=1, Ny=1, Nz=1)
├── tracers: (:T, :S)
├── closure: HorizontallyCurvilinearAnisotropicBiharmonicDiffusivity{Float64, Nothing, NamedTuple{(:T, :S), Tuple{Float64, Float64}}, Nothing}
├── buoyancy: Buoyancy{SeawaterBuoyancy{Float64, LinearEquationOfState{Float64}, Nothing, Nothing}, Oceananigans.Grids.ZDirection}
└── coriolis: Nothing

julia> model.grid.Hx, model.grid.Hy, model.grid.Hz
(1, 1, 1)
```

Reply:
Ah. There's a small issue because merely querying `inflate_halo_size` throws that warning, even if we don't actually replace the grid. I'll fix that.

Reply:
>Because HorizontallyCurvilinearAnisotropicBiharmonicDiffusivity() requires 2 halos?

precisely